python/private/bzlmod/pip.bzl

```diff
@@ -379,6 +466,22 @@ def _pip_impl(module_ctx):
 
 def _pip_parse_ext_attrs():
     attrs = dict({
+        "experimental_index_url": attr.string(
```

Comment:
There are two things in `pip`:
* `--index-url` - this is the default index
* `--extra-index-urls` - these are the extra indexes that would be used in addition to the one above. I am not supporting this usecase as it makes the code more complex.

I see how the `ex` word here may make people think about `--extra-index-urls` argument. Shall I add something in the docs?

Reply:
IME it's very rare to have a full private index[^1]. Instead, the common case is to use the public PyPI.org as one index and then add the private index via `--extra-index-urls`.

So I think _only_ supporting `--index-url` will not help many people :worried:.

[^1]: Though from a security perspective this is the best thing to do: have a mirror of public pypi where you allowlist only the packages that people need and denylist everything else. The problem is that's a lot of overhead and it's seen as "not worth it" :disappointed:

Reply:
If we add an extra index it would be good to know what packages we should fetch from there because we would get 404 errors at best in the pypi public and wrong results at worst.

For artifactory, you can create virtual repos that would be a union of pypi index and the internal one. The logic of what comes from which index would be handled by the index. I'm not sure about other indexes though.

I am thinking that even though this might not be the best API now, I don't see how we could get it better with what we have now. Maybe in the long run we could use uv/pip install dry-run report to get an idea about which package comes from which index, but I would love to postpone this to a future PR. @philsc has done greate work to get the install reports integrated in rules_python, so maybe we can collaborate to come up with something better some time later? :)

Reply:
❤️ thank you!

I'll say that getting 404s in public pypi is already how `pip` works with `--extra-index-url` so I'm not sure if that's a dealbreaker. `pip` first tries to find the package at `--index-url` and if it fails there it goes to `--extra-index-url`s in order. (This is a Bad Thing :tm: because of exactly what you mentioned[^1], but sadly that's how it currently works.)

> For artifactory, you can create virtual repos that would be a union of pypi index and the internal one.

Ooooo I didn't know that, that's a nice feature! Looks like Google Artifact Registry has that too - I'll have to tell our SWOps people.

[^1]: Example: Malicious actor finds out that we have a package `foo-bar-company` on our private index and then namesquats a bad package with the same name on the public index. 👎

Reply:
I think this convo can be resolved? I'll mark it as such